x/concentrated-liquidity/swaps_test.go

```diff
@@ -158,3 +153,145 @@ func (s *KeeperTestSuite) TestCLPoolSimpleSwapExactAmountIn() {
+			param: param{
+				tokenOut:         sdk.NewCoin("eth", sdk.NewInt(13370)),
+				tokenInDenom:     "eth",
+				tokenInMaxAmount: types.UpperPriceLimit.RoundInt(),
+			},
+			expectErr: "cannot trade same denomination in and out",
+		},
+		{
+			name: "unknown in denom",
+			param: param{
+				tokenOut:         sdk.NewCoin("etha", sdk.NewInt(13370)),
+				tokenInDenom:     "eth",
+				tokenInMaxAmount: types.UpperPriceLimit.RoundInt(),
+			},
+			expectErr: "does not match any asset in pool",
+		},
+		{
+			name: "unknown out denom",
+			param: param{
+				tokenOut:         sdk.NewCoin("eth", sdk.NewInt(13370)),
+				tokenInDenom:     "etha",
+				tokenInMaxAmount: types.UpperPriceLimit.RoundInt(),
+			},
+			expectErr: "does not match any asset in pool",
+		},
+	}
+
+	for _, test := range tests {
+		s.Run(test.name, func() {
+			// Init suite for each test.
+			s.Setup()
+
+			// Create a usdc - eth concentrated liquidity pool
+			pool, err := s.App.ConcentratedLiquidityKeeper.CreateNewConcentratedLiquidityPool(s.Ctx, 1, "eth", "usdc", DefaultCurrSqrtPrice, DefaultCurrTick)
+			s.Require().NoError(err)
+
+			// Check the test case to see if we are swapping asset0 for asset1 or vice versa
+			asset0 := pool.GetToken0()
+			zeroForOne := test.param.tokenOut.Denom == asset0
+
+			// Fund the test account with usdc and eth, then create a default position to the pool created earlier
+			s.SetupPosition(1)
+
+			// Note spot price and gas used prior to swap
+			spotPriceBefore := pool.GetCurrentSqrtPrice().Power(2)
+			prevGasConsumed := s.Ctx.GasMeter().GasConsumed()
+
+			// Execute the swap directed in the test case
+			tokenIn, err := s.App.ConcentratedLiquidityKeeper.SwapExactAmountOut(s.Ctx, s.TestAccs[0], pool.(swaproutertypes.PoolI), test.param.tokenInDenom, test.param.tokenInMaxAmount, test.param.tokenOut, DefaultZeroSwapFee)
+
+			if test.expectErr != "" {
+				s.Require().ErrorContains(err, test.expectErr)
+			} else {
+				s.Require().NoError(err)
+				s.Require().Equal(test.param.expectedTokenIn.String(), tokenIn.String())
+
+				gasConsumedForSwap := s.Ctx.GasMeter().GasConsumed() - prevGasConsumed
+				// Check that we consume enough gas that a CL pool swap warrants
+				// We consume `types.GasFeeForSwap` directly, so the extra I/O operation mean we end up consuming more.
+				s.Require().Greater(gasConsumedForSwap, uint64(cltypes.ConcentratedGasFeeForSwap))
+
+				// Assert events
+				s.AssertEventEmitted(s.Ctx, cltypes.TypeEvtTokenSwapped, 1)
+
+				// Retrieve pool again post swap
+				pool, err = s.App.ConcentratedLiquidityKeeper.GetPoolById(s.Ctx, pool.GetId())
+				s.Require().NoError(err)
+
+				spotPriceAfter := pool.GetCurrentSqrtPrice().Power(2)
+
+				// Ratio of the token out should be between the before spot price and after spot price.
+				tradeAvgPrice := tokenIn.ToDec().Quo(test.param.tokenOut.Amount.ToDec())
+				if !zeroForOne {
+					tradeAvgPrice = sdk.OneDec().Quo(tradeAvgPrice)
+				}
```

Comment:
nit: there is another check for `zeroForOne` right below, we can move this logic there